From d49b79e7f0e121e331861395504f60e6281fe02f Mon Sep 17 00:00:00 2001
From: Yuan Can <yuancan@huawei.com>
Date: Wed, 23 Aug 2023 15:47:40 +0800
Subject: [PATCH] mm: Fix static hugepages decrease failed on HBM nodes

With all HBM nodes set to movable nodes, if static hugepages allocated from
HBM node with hugetlb_optimize_vmemmap turns on, decreasing the number of
hugepages can fail with the following call trace:

   sh: page allocation failure: order:0, mode:0x210cc0(GFP_KERNEL|__GFP_NORETRY|__GFP_THISNODE)
   CPU: 5 PID: 867 Comm: sh Tainted: G        W         5.10.0+ #118
   Hardware name: linux,dummy-virt (DT)
   Call trace:
    dump_backtrace+0x0/0x1e0
    show_stack+0x2c/0x38
    dump_stack+0xc8/0x100
    warn_alloc+0xf0/0x168
    __alloc_pages+0xd30/0xe10
    vmemmap_remap_alloc+0xc4/0x1d0
    hugetlb_vmemmap_alloc+0x60/0xa8
    __update_and_free_page+0x2c/0x278
    update_and_free_page+0x40/0x90
    update_and_free_pages_bulk+0x44/0x70
    set_max_huge_pages+0x1c8/0x408
    nr_hugepages_store_common+0xd0/0x178
    nr_hugepages_store+0x38/0x48
    kobj_attr_store+0x14/0x28
    sysfs_kf_write+0x58/0x80
    kernfs_fop_write_iter+0x150/0x1c8
    new_sync_write+0xfc/0x1a0
    vfs_write+0x248/0x2a8
    ksys_write+0x74/0xf8

The reason is that vmemmap_remap_alloc tries to alloc vmemmap pages from
HBM node, with all HBM nodes set to movable nodes, HBM nodes have only
movable zone, and GFP_KERNEL cannot alloc from movable zone, __GFP_THISNODE
forbit it from falling back to DDR node.
As the struct pages is expected to alloc from DDR node, remove
__GFP_THISNODE so it can fall back to DDR node to fix this.

Signed-off-by: Yuan Can <yuancan@huawei.com>
---
 mm/hugetlb_vmemmap.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/hugetlb_vmemmap.c b/mm/hugetlb_vmemmap.c
index 91a9e3626..8d1042ace 100644
--- a/mm/hugetlb_vmemmap.c
+++ b/mm/hugetlb_vmemmap.c
@@ -254,7 +254,7 @@ int hugetlb_vmemmap_alloc(struct hstate *h, struct page *head)
 	 * discarded vmemmap pages must be allocated and remapping.
 	 */
 	ret = vmemmap_remap_alloc(vmemmap_addr, vmemmap_end, vmemmap_reuse,
-				  GFP_KERNEL | __GFP_NORETRY | __GFP_THISNODE);
+				  GFP_KERNEL | __GFP_NORETRY);
 	if (!ret) {
 		ClearHPageVmemmapOptimized(head);
 		static_branch_dec(&hugetlb_optimize_vmemmap_key);
-- 
2.17.1

